sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/ReadFromKafkaDoFn.java

```diff
@@ -311,17 +324,26 @@ public OffsetRangeTracker restrictionTracker(
     if (restriction.getTo() < Long.MAX_VALUE) {
       return new OffsetRangeTracker(restriction);
     }
-    Map<String, Object> updatedConsumerConfig =
-        overrideBootstrapServersConfig(consumerConfig, kafkaSourceDescriptor);
-    KafkaLatestOffsetEstimator offsetPoller =
-        new KafkaLatestOffsetEstimator(
-            consumerFactoryFn.apply(
```

Comment:
factor bad / pool good ?

Reply:
Specifically per-topic-partition consumer pool? I'm inexpert but seems like this is the right granularity to be reusing objects/connections.

Reply:
I may not follow why we need a pool. This proposal is to reuse the same consumer / estimator for the same topic partition over and over -- so not more than 1 connection.

The intent was already to use memoization, so it's clearly just an estimate of the backlog.

Reply:
We saw ~5 rather than ~1 didn't we? I'm in too many chats on this :-). At heart, I just don't like the code approach of counting on memoizing a result value as the way we avoid creating a connection, and I prefer having the automatic creation take place at a higher level in the call stack, ideally at the DoFn level. I see that SDF and UnboundedSource have the difference that the topic partition is statically determined vs incoming elements. TBH I'd not expect a major architectural difference. To be continued!